From 1066e1f22ea0a3c4de38e04d02320588b2eed33f Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:10:53 -0400
Subject: [powerpc] tracing: TRACE_WITH_FRAME_BUFFER creates invalid stack frames

Message-id: <1410545655-205645-425-git-send-email-dzickus@redhat.com>
Patchwork-id: 94548
O-Subject: [RHEL7 PATCH 424/626] powerpc/tracing: TRACE_WITH_FRAME_BUFFER creates invalid stack frames
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit d51959d70ffc55d1c829e881a6121e6fbbfb29af
Author: Anton Blanchard <anton@samba.org>
Date:   Mon Mar 10 12:51:58 2014 +1100

    powerpc/tracing: TRACE_WITH_FRAME_BUFFER creates invalid stack frames

    TRACE_WITH_FRAME_BUFFER creates 32 byte stack frames. On ppc64
    ABIv1 this is too small and a callee could corrupt the stack by
    writing to the parameter save area (starting at offset 48).

    Signed-off-by: Anton Blanchard <anton@samba.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/irqflags.h b/arch/powerpc/include/asm/irqflags.h
index f62c056..e20eb95 100644
--- a/arch/powerpc/include/asm/irqflags.h
+++ b/arch/powerpc/include/asm/irqflags.h
@@ -20,9 +20,9 @@
  */
 #define TRACE_WITH_FRAME_BUFFER(func)  \
  mflr r0;    \
- stdu r1, -32(r1);   \
+ stdu r1, -STACK_FRAME_OVERHEAD(r1); \
  std r0, 16(r1);   \
- stdu r1, -32(r1);   \
+ stdu r1, -STACK_FRAME_OVERHEAD(r1); \
  bl func;    \
  ld r1, 0(r1);   \
  ld r1, 0(r1);
-- 
1.7.1